docs/design/datacontracts/datacontracts_design.md

```diff
@@ -0,0 +1,317 @@
+
+### Algorithmic contracts
+Algorithmic contracts define how to process a given set of data structures to produce useful results. These are effectively code snippets which utilize the abstracted data structures provided by Data Structure Definition Contracts and Global Value Contract to produce useful output about a given program. Descriptions of these contracts may refer to functionality provided by other contracts to do their work. The algorithms provided in these contracts are designed to operate given the ability to read various primitive types and defined data structures from the process memory space, as well as perform general purpose computation.
+
+It is entirely reasonable for an algorithmic contract to have multiple entrypoints which take different inputs. For example imagine a contract which provides information about a `MethodTable`. It may provide the an api to get the `BaseSize` of a `MethodTable`, and an api to get the `DynamicTypeID` of a `MethodTable`. However, while the set of contracts which describe an older version of .NET may provide a means by which the `DynamicTypeID` may be acquired for a `MethodTable`, a newer runtime may not have that concept. In such a case, it is very reasonable to define that the `GetDynamicTypeID` api portion of that contract is defined to simply `throw new NotSupportedException();`
+
+For simplicity, as it can be expected that all developers who work on the .NET runtime understand C# to a fair degree, it is preferred that the algorithms be defined in C#, or at least psuedocode that looks like C#. It is also condsidered entirely permissable to refer to other specifications if the algorithm is a general purpose one which is well defined by the OS or some other body. (For example, it is expected that the unwinding algorithms will be defined by references into either the DWARF spec, or various Windows Unwind specifications.)
+
+For working with data from the target process/other contracts, the following C# interface is intended to be used within the algorithmic descriptions:
+
+Best practice is to either write the algorithm in C# like psuedocode working on top of the [C# style api](contract_csharp_api_design.cs) or by reference to specifications which are not co-developed with the runtime, such as OS/architecture specifications. Within the contract algorithm specification, the intention is that all interesting api work is done by using an instance of the `Target` class.
+
+## Arrangement of contract specifications in the repo
+
+Specs shall be stored in the repo in a set of directories. `docs/design/datacontracts` Each one of them shall be a seperate markdown file named with the name of contract. `docs/design/datacontracts/datalayout/<contract_name>.md` Every version of each contract shall be located in the same file to facilitate understanding how variations between different contracts work.
```

Comment:
I think it might be useful to have a machine-readable header to the contract specs.   This will allow some kind of cross-validation:
1. That the reader implements all known contracts, doesn't misspell any contract names, or miss a version.
2. That the target runtime descriptor includes only known names/versions and doesn't misspell any names.

Additionally for globals/offsets we could use the machine-readable spec to either generate or validate the implementations in the reader.

For example using [YAML frontmatter metadata](https://docs.github.com/en/contributing/writing-for-github-docs/using-yaml-frontmatter#yaml-frontmatter-values).  And then leave the markdown purely for descriptive text.  Something like:

```markdown
---
contract_name: "XYZContract"
versions: [1, 2, 5]
contract_type: "global"
globals:
  - name: SomeGlobal
     type: int32
     value: 1
  - name: SomeOtherGlobal
     type: int8
     value: 0
version_1:
  - globals:
     - name: SomeGlobal
       value: 0
version_5:
  - globals:
     - name: ThirdGlobal
        type: int8
        value: 5
---

# Contract XYZ_Contract

Descriptive text

## Version 1

What's interesting about Version 1

## Version 2

This is also interesting because SomeGlobal changed values

## Version 5

This version changed SomeGlobal back to its original value, but also added ThirdGlobal

```

We dont' necessarily need to build the tooling on day 1, but it would be nice to have the option for the future.  And it's not really any more work for contract authors to follow a slightly constrained format.

Reply:
I have no objections, and in fact the formatting of the data layout/globals stuff is something I'm not focusing on too much today. I'd like to hear other perspectives though. @jkotas, @noahfalk?

Reply:
What would be a real-world example of `SomeGlobal`? I am not sure how to think about it. I would expect that each contract version can be used with multiple different values of SomeGlobal.

The non-YAML part of your example that has descriptions of the significant changes in each version looks good.

Reply:
Mostly this seems good to me 👍 

One caveat is I'm still hoping we don't need to have different types of contracts so `contract_type: "global"` wouldn't exist. Instead I was imagining that any contract is allowed to include globals as well as other elements.

Reply:
> What would be a real-world example of `SomeGlobal`? I am not sure how to think about it. I would expect that each contract version can be used with multiple different values of SomeGlobal.

yea, I'm also not sure about this part.  

One thing that @davidwrighton has talked about is that for algorithm contracts new versions are allowed to throw if a certain method is no longer relevant in a newer version of a contract. 

I'm not sure if it makes sense for global access to throw if a newer version of a contract makes some global obsolete or if it is better for obsolete globals to be specified in a newer version of a contract to have some default constant value.  I could imagine both to be useful in different situations.

> The non-YAML part of your example that has descriptions of the significant changes in each version looks good.

I think the absolute minimal YAML part of my proposal is just something like this:

```yaml
---
contract: "XYZContract"
versions: [1, 2, 3]
---
```

I'm absolutely 100% certain I will check in typos in the reader implementation if there isn't some tooling to spellcheck the valid contract names.